File: runtime/java/units/wasi/fd_close.java
// Close a file fd (its FileChannel) and drop it from the table. Stdio and
// directory fds carry no OS handle to close. Fds are never reused after close
// (ADR-14).
int wasi_fd_close(int fd) {
    if (!fds.containsKey(fd)) {
        return WASI_BADF;
    }
    Object e = fds.remove(fd);
    meta.remove(fd);
    if (e instanceof Handle) {
        try {
            ((Handle) e).ch.close();
        } catch (java.io.IOException ex) {
            return WASI_IO;
        }
    }
    return WASI_OK;
}

File: runtime/go/units/wasi/fd_close.go
func (w *WASI) wasi_fd_close(fd uint32) uint32 {
    entry, present := w.fds[fd]
    if !present {
        return wasiBadf
    }
    delete(w.fds, fd)
    if f, ok := entry.(*os.File); ok && !w.isStdio(f) {
        f.Close()
    }
    // Fds are never reused after close; a *wasiDir has no OS handle.
    return wasiOk
}
